traversal/walk.go

```diff
@@ -142,192 +157,267 @@ func (prog Progress) WalkLocal(n datamodel.Node, fn VisitFn) error {
 // An AdvVisitFn is used instead of a VisitFn, so that the reason can be provided.
 func (prog Progress) WalkAdv(n datamodel.Node, s selector.Selector, fn AdvVisitFn) error {
 	prog.init()
-	return prog.walkAdv(n, s, fn)
+	return prog.walkBlock(n, s, fn)
+}
+
+// walkBlock anchors a walk at the beginning of the traversal and at the
+// beginning of each new link traversed. This allows us to do a preload phase if
+// we have a preloader configured.
+func (prog Progress) walkBlock(n datamodel.Node, s selector.Selector, visitFn AdvVisitFn) error {
+	ph := phaseTraverse
+	var budget *Budget
+
+	if prog.Cfg.Preloader != nil {
+		ph = phasePreload
+		budget = prog.Budget.Clone()
+	}
+
+	// First pass.
+	err := prog.walkAdv(ph, n, s, visitFn)
+	if err != nil && (ph != phasePreload || !errors.Is(&ErrBudgetExceeded{}, err)) {
+		return err
+	}
+
+	if ph == phasePreload {
+		// First past was a preload; now do the _real_ pass.
+		prog.Budget = budget // reset
+		return prog.walkAdv(phaseTraverse, n, s, visitFn)
+	}
+
+	return nil
+}
+
+// walkAdv is the main recursive walk function, called to iterate through
+// recursive nodes (root node, maps, lists and new link root nodes).
+func (prog Progress) walkAdv(ph phase, n datamodel.Node, s selector.Selector, visitFn AdvVisitFn) error {
+	if err := prog.checkNodeBudget(); err != nil {
+		return err
+	}
+
+	// If we need to interpret this node in an alternative form, reify and replace.
+	if rn, rs, err := prog.reify(n, s); err != nil {
+		return err
+	} else if rn != nil {
+		n = rn
+		s = rs
+	}
+
+	// Call the visit function if necessary.
+	if err := prog.visit(ph, n, s, visitFn); err != nil {
+		return err
+	}
+
+	// If we're handling scalars (e.g. not maps and lists) we can return now.
+	switch n.Kind() {
+	case datamodel.Kind_Map, datamodel.Kind_List: // continue
+	default:
+		return nil
+	}
+
+	// For maps and lists: recurse (in one of two ways, depending on if the selector also states specific interests).
+
+	haveStartAtPath := prog.Cfg.StartAtPath.Len() > 0
+	var reachedStartAtPath bool
+	recurse := func(v datamodel.Node, ps datamodel.PathSegment) error {
```

Comment:
I believe the overhead of the function itself is fairly minimal, but it will obviously hold on to any variables closed over. In this case there's 2 booleans, a `progress` and the `Node` and `PathSegment` arguments.

But, I think it's only the closure reference and one of the booleans that are new in what would be held onto in the recursive call stack. Unless there's something special I'm not aware of with Go functions used in this way then I believe the overhead is negligible compared to what else we're holding on to through the recursion.